modules/artifact/src/main/java/org/mule/runtime/module/artifact/api/descriptor/ClassLoaderModel.java

```diff
@@ -157,10 +182,34 @@ public ClassLoaderModelBuilder exportingResources(Set<String> resources) {
       return this;
     }
 
+    /**
+     * Indicates which Java packages are loaded from the artifact itself, even if some other artifact in the region exports it.
+     *
+     * @param packages packages to not import. Non null.
+     * @return same builder instance.
+     */
+    public ClassLoaderModelBuilder notImportingPackages(Set<String> packages) {
+      checkArgument(packages != null, "packages cannot be null");
+      this.notImportedPackages.addAll(packages);
+      return this;
+    }
+
+    /**
+     * Indicates which resources are loaded from the artifact itself, even if some other artifact in the region exports it.
+     *
+     * @param resources resources to not import. Non null.
+     * @return same builder instance.
+     */
+    public ClassLoaderModelBuilder notImportingResources(Set<String> resources) {
+      checkArgument(resources != null, "resources cannot be null");
```

Comment:
On the Java package side, you can check if the package is exported in the artifact's API and transitively determine the closure, but to review which resources can or not be consumed locally requires more code analysis.